### ISSUE TITLE
About the generated labels

### ISSUE THREAD

**Issue by Zhengtq** [lorenmt/maxl#5]
I notice that the pseudo-label generated by the 'LabelGenerator' is a so called masked softmax probability which is not one-hot labels as commonly used in multi-class cross entropy loss. In my experiment, I transfer the maxl method to my task(binary classification, TF implementation) and find that each class probability in the generated labels is pretty close to each other. So every class logits in the aux_logtis that generated by the auxiliary branch would be very close. Just like below(binary classification, class zero has 5 aux_classes, class one has 15 aux_class):

A sample's pseudo-label of class zero(after masked softmax):
`
[0.04 0.13 0.3  0.26 0.27 0.   0.   0.   0.   0.   0.   0.   0.   0.
 0.   0.   0.   0.   0.   0.  ]`
predicted aux_logits of the corresponding sample(after softmax):
`
[ 5.1   5.18  5.17  5.16  5.14 -1.98 -1.95 -2.25 -1.79 -1.25 -1.75 -2.05
 -1.78 -1.68 -2.   -2.09 -1.65 -1.65 -1.8  -1.5 ]
`

A sample's pseudo-label of class one(after masked softmax):
`
[0.   0.   0.   0.   0.   0.01 0.12 0.02 0.1  0.06 0.12 0.25 0.02 0.1
 0.01 0.01 0.06 0.01 0.04 0.06]
`
predicted aux_logits of the corresponding sample(after softmax):
`
[-9.61 -9.96 -9.5  -9.6  -8.82  3.43  3.56  3.4   3.57  3.49  3.45  3.32
  3.44  3.54  3.45  3.45  3.59  3.48  3.49  3.61]
`

As you can see, because of the pseudo-label is not one-hot, the aux_branch could be easily collapsed to produce similarly logits.  And obviously the model entropy is not designed to solve this kind of model collapse.(As I understand it, Model entropy in you paper is designed to make logtis more diverse in each aux_class.)
So could you point out what causes this problem and how to solve it.

**Comment by lorenmt** [comment]
You can consider those soft labels generated by the label generators are in similar to the one from a teacher-student model, aiming for more accurate representation compared to one-hot descriptor. 

You may use a max function to produce a one-hot vector in your loss function, as max function is still differentiable. But from the experiments, we found out that these soft labels can produce a better performance compared to the hard labels.

Or simply, you can tune down the hyper-parameter weighting for model entropy. If such entropy is too large, the generator would then entirely collapse to produce the same class probability for any training data.

Hope that helps.

**Comment by Zhengtq** [comment]
Thanks for providing more insight about your work.

**Comment by Zhengtq** [comment]
As you say:

> You can consider those soft labels generated by the label generators are in similar to the one from a teacher-student model, aiming for more accurate representation compared to one-hot descriptor.

Could I use cosine similarity to measure the distribution similarity between the soft labels generated and the aux_logits produced by aux_branch?

**Comment by lorenmt** [comment]
Yes, you can. But I don't think it would be an effective approach to treat it as a loss and do back-prop.

**Comment by Zhengtq** [comment]
Thanks for you reply. I would discuss with your if more insights are being got.
